ZFS fix to make xattr=sa logging to ZIL on create/remove/update.

jsai20 opened the pull request:
As such, there are no specific synchronous semantics defined for
the xattrs. But for xattr=on, it does log to ZIL and zil_commit() is
done, if sync=always is set on dataset. This provides sync semantics
for xattr=on with sync=always set on dataset.
For the xattr=sa implementation, it doesn't log to ZIL, so, even with
sync=always, xattrs are not guaranteed to be synced before xattr call
returns to caller. so xattr can be lost if system crash happens, before
txg carrying xattr transaction is synced.

This change makes xattr=sa logging to ZIL on xattr create/remove/update
and xattrs are synced to ZIL (zil_commit() done) for sync=always.
This make xattr=sa behavior similar to xattr=on.

This could also provide basic framework to support implementing sync
semantics at file level for xattr=sa.

Signed-off-by: Jitendra Patidar <jitendra.patidar@nutanix.com>
Closes #8768

ahrens commented:
Can you clarify whether we are talking about when synchronous semantics are requested vs not?  How do we request those semantics?  I think that we only log these operations do disk if sync=always or when we later do another (unrelated) sync operation, is that right?  

It looks like with the default of xattr=on, these would be logged via zpl_xattr_set_dir()'s calls to zfs_create(), zpl_write_common(), etc.  These would do zfs_log_create() but not zil_commit().  Are those semantics correct?

I think that with xattr=sa, we are not creating the itx at all (i.e. not calling zfs_log_create()).  So there is no way for these operations to be logged.  That does indeed seem like a bug.

behlendorf commented:
> Can you clarify whether we are talking about when synchronous semantics are requested vs not? How do we request those semantics? 

Good question.  Unlike illumos and FreeBSD, the `setxattr(2)` system call is used to atomically manipulate xattr's on Linux.  The expected behavior is described in the `attr(5)` man page and it differs slightly between filesystems.  But it does not explicitly state if `setxattr(2)` is expected to behave synchronously.  However, since they're implemented as a system call and one of the primary uses for xattrs is to store ACLs and selinux contexts I think making them synchronous would be prudent.

> I think that with xattr=sa, we are not creating the itx at all (i.e. not calling zfs_log_create()). So there is no way for these operations to be logged. That does indeed seem like a bug.

Adding support to log these wasn't included in the original implementation, and we never managed to revisit this and add the missing support.  Which is definitely a bug.

A new ZIL log record type does seem like the correct way to handle this, but we'll need to register a new feature flag for this.  Otherwise, replaying a log with these entries on another OpenZFS platform will result in a crash.

jsai20 commented:
> > Can you clarify whether we are talking about when synchronous semantics are requested vs not? How do we request those semantics?
> 
> Good question. Unlike illumos and FreeBSD, the `setxattr(2)` system call is used to atomically manipulate xattr's on Linux. The expected behavior is described in the `attr(5)` man page and it differs slightly between filesystems. But it does not explicitly state if `setxattr(2)` is expected to behave synchronously. However, since they're implemented as a system call and one of the primary uses for xattrs is to store ACLs and selinux contexts I think making them synchronous would be prudent.
> 
> > I think that with xattr=sa, we are not creating the itx at all (i.e. not calling zfs_log_create()). So there is no way for these operations to be logged. That does indeed seem like a bug.
> 
> Adding support to log these wasn't included in the original implementation, and we never managed to revisit this and add the missing support. Which is definitely a bug.
> 
> A new ZIL log record type does seem like the correct way to handle this, but we'll need to register a new feature flag for this. Otherwise, replaying a log with these entries on another OpenZFS platform will result in a crash.

Thanks Matthew and Brain for the review comments.  yes,  exact semantics (sync or async) required for the operation are not very clear.   For xattr=on, it does log the xattr write/update operation via zpl_write_common() path as Metthew mentioned and with sync=always, log would be committed to disk.  But for xattr=sa, it doesn't log the operation. So, for xattr=sa, as well, similarly, we can log the record and commit (zil_commit) to disk as part of operation if sync=always.

Thanks Brian, I would check the feature flag part. I am new to zfs and still learning it.

ahrens commented:
> since they're implemented as a system call and one of the primary uses for xattrs is to store ACLs and selinux contexts I think making them synchronous would be prudent.

Do you think that we should also change the xattr=on behavior to be always sync?

In either case, I think that like all other ZFS operations (AFAIK), xattr operations (whether xattr=on or xattr=sa) should be async by default (i.e. not persisted to disk immediately via zil_commit()).

jsai20 commented:
> > since they're implemented as a system call and one of the primary uses for xattrs is to store ACLs and selinux contexts I think making them synchronous would be prudent.
> 
> Do you think that we should also change the xattr=on behavior to be always sync?
> 
> In either case, I think that like all other ZFS operations (AFAIK), xattr operations (whether xattr=on or xattr=sa) should be async by default (i.e. not persisted to disk immediately via zil_commit()).

For xattr=on,  xattr's are synced to zil via zpl_xattr_set_dir()->zpl_write_common()->zpl_write_common_iovec()->zfs_write() { 
  ....
  if (ioflag & (FSYNC | FDSYNC) ||
            zfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS)
                zil_commit(zilog, zp->z_id);
...
}
As ioflag seems to be alwyas passed as 0 for xattr write. So, zil_commit is only done if dataset property "sync" is set to "always".
xattr=on should always do sync to disk, which means, zil_commit() should always be done, without any checks. So, I am not really sure about it.

In my diff, which enables xattr=sa logging to zil, for now I have kept xattr=sa behavior same as xattr=on, means do zil_commit() only if dataset property "sync" is set to "always". So, xattr=sa attributes doesn't lost on node crash similar to xattr=on, but it's guaranteed only when sync=always is set.

I think, decision about "xatt=on" and "xattr=sa" should always do zil_commit() needs more involved discussion and not much related to diff here. Appropriate diff/pull request for the same can be opened separately, as discussed and decided.

I have updated diff with feature flag change. Please have a look and provide feedback on same.

ahrens commented:
> xattr=on should always do sync to disk, which means, zil_commit() should always be done, without any checks

Why should xattr=on always sync to disk immediately?

> I think, decision about "xatt=on" and "xattr=sa" should always do zil_commit() needs more involved discussion and not much related to diff here.

I agree that it isn't related to this PR, so we can proceed with `xattr=sa` having the same semantics as `xattr=on` currently does.

jsai20 commented:
> > xattr=on should always do sync to disk, which means, zil_commit() should always be done, without any checks
> 
> Why should xattr=on always sync to disk immediately?
I meant, I am not very sure, either it should be always sync or not. I can't find any specific reference to it.
> 
> > I think, decision about "xatt=on" and "xattr=sa" should always do zil_commit() needs more involved discussion and not much related to diff here.
> 
> I agree that it isn't related to this PR, so we can proceed with `xattr=sa` having the same semantics as `xattr=on` currently does.
Thanks.

jsai20 commented:
> > xattr=on should always do sync to disk, which means, zil_commit() should always be done, without any checks
> 
> Why should xattr=on always sync to disk immediately?
> 

I meant, I am not very sure, either it should be always sync or not. I can't find any specific reference to it.

> > I think, decision about "xatt=on" and "xattr=sa" should always do zil_commit() needs more involved discussion and not much related to diff here.
> 
> I agree that it isn't related to this PR, so we can proceed with `xattr=sa` having the same semantics as `xattr=on` currently does.

Thanks.

snajpa commented:
When talking new feature flags - there's an ultimate trade-off between these two options with the current design:
1. making the pools used on Linux r/o on other platforms permanently by increasing the feature count on first use
2. increment/decrement the feature on every use of the tx types - but the functions doing that are actually supposed to be called in a transaction group syncing context.

It is explained a bit better in the comment in the OverlayFS PR #9414 
 https://github.com/zfsonlinux/zfs/pull/9414/files#r340373675

I don't understand the need for those functions to be called in the syncing context fully, but I suppose it is to be sure that the incr/decr actually lands on disk by the end of the txg.

So, might I suggest adding new txtypes for incr/decr of ZFS pool features? That way, we could incr/decr the features no matter the context; the only trade off here is that on import, to check for the features used by the pool, we'd have to walk the IL. I'm not sure whether this can be done in a flat way for the whole pool, or whether we have to walk the dataset structure and then walk the associated IL records...

@behlendorf @ahrens perhaps you would know more and could advise on the matter?

I wouldn't mind doing the work, but I'd be glad for some guidance :)

ahrens commented on module/zfs/zfs_replay.c:
```diff
@@ -860,6 +863,89 @@ zfs_replay_setattr(void *arg1, void *arg2, boolean_t byteswap)
+	size_t size;
+	int error = 0;
+
+	ASSERT(spa_feature_is_enabled(zfsvfs->z_os->os_spa,
```
I think this should be `spa_feature_is_active`

jsai20 replied:
yes. Changed.

ahrens commented on module/zcommon/zfeature_common.c:
```diff
@@ -553,6 +553,17 @@ zpool_feature_init(void)
+	    "Zil sa xattr.",
+	    ZFEATURE_FLAG_READONLY_COMPAT | ZFEATURE_FLAG_PER_DATASET,
+	    ZFEATURE_TYPE_BOOLEAN, zilsaxattr_deps);
+	}
```
I don't see where this gets activated.  I think this would have to be ZFEATURE_FLAG_ACTIVATE_ON_ENABLE (and not ZFEATURE_FLAG_PER_DATASET).

jsai20 replied:
yes. Changed it.

ahrens commented on module/zfs/zfs_sa.c:
```diff
@@ -243,6 +248,11 @@ zfs_sa_set_xattr(znode_t *zp)
 
+	zilog = zfsvfs->z_log;
+	if (spa_feature_is_enabled(zfsvfs->z_os->os_spa,
+	    SPA_FEATURE_ZILSAXATTR) && zfs_zil_saxattr_enabled) {
```
since you can't activate it here, I think this needs to be  `spa_feature_is_active()`

jsai20 replied:
yes.

ahrens commented:
@snajpa What you are suggesting would most likely be implemented something like this:

When we do a read-write import of the pool, when we walk all the ZIL's to claim their blocks, we also check if there are any unrecognized record types.  If so, we fail the import.  One downside with this is that the record types are just an enum, so all implementations of ZFS must agree about what each value means.

Alternatively, we could add a single feature flag now which is always activated (refcount=1), and add a single new ZIL record type.  This new record type would specify which other feature flag must be supported in order to import the pool read-write.  The other feature flag would be specified by fully-qualified name, so it wouldn't have the downside mentioned above.  But it doesn't solve the backwards-compatibility for the first thing that needs it.

Personally, I think we should err on the side of simplicity.  We already have the feature flag infrastructure.  We shouldn't shy away from using it to introduce new feature flags and accept the fact that older/different implementations won't be able to access the pool (at least not writeable).  But I could be convinced otherwise depending on the specifics.  IMO, the problem this PR is solving is not that acute, so folks desiring compatibility can leave the feature disabled with little impact.

ahrens commented:
It looks like maybe we decided on adding a new on-disk feature flag for this, but the code review process stalled out.  @jsai20 if you were to rebase this, would it be ready for final code review and integration, as far as you know?

jsai20 commented:
> It looks like maybe we decided on adding a new on-disk feature flag for this, but the code review process stalled out. @jsai20 if you were to rebase this, would it be ready for final code review and integration, as far as you know?

Thanks Matt. ya, code review was stalled after around conversation on new on-disk feature flag.
I would just recollect, rebase the change and update the diff.

freqlabs commented:
This will need to be applied on the FreeBSD side as well, and the test updated to use helpers from libtest for working with tunables and xattrs.

freqlabs commented on tests/zfs-tests/tests/functional/slog/slog_016_pos.ksh:
```diff
@@ -0,0 +1,138 @@
+{
+	saxattrzil=$1
+
+	echo $saxattrzil > /sys/module/zfs/parameters/zfs_zil_saxattr_enabled
```
Add the tunable to tests/zfs-tests/include/tunables.cfg and use set_tunable32 here.

jsai20 replied:
Done

freqlabs commented on tests/zfs-tests/tests/functional/slog/slog_016_pos.ksh:
```diff
@@ -0,0 +1,138 @@
+	#
+	for i in $(seq $NFILES); do
+		log_must mkdir /$TESTPOOL/$TESTFS/xattr.d.$i
+		log_must attr -qs test -V test /$TESTPOOL/$TESTFS/xattr.d.$i
```
set_xattr name value path

jsai20 replied:
Done

freqlabs commented on tests/zfs-tests/tests/functional/slog/slog_016_pos.ksh:
```diff
@@ -0,0 +1,138 @@
+	#
+	for i in $(seq $NFILES); do
+		if [ $saxattrzil -eq 0 ]; then
+			log_mustnot attr -g test /$TESTPOOL/$TESTFS/xattr.d.$i
```
get_xattr name path

jsai20 replied:
Done

freqlabs commented on module/os/linux/zfs/zpl_xattr.c:
```diff
@@ -578,7 +578,7 @@ zpl_xattr_set_sa(struct inode *ip, const char *name, const void *value,
 	 */
 	if (error == 0)
-		error = -zfs_sa_set_xattr(zp);
+		error = -zfs_sa_set_xattr(zp, name, value, size);
```
This change is also needed in `module/os/freebsd/zfs/zfs_vnops_os.c`

jsai20 replied:
Done

jsai20 commented:
> This will need to be applied on the FreeBSD side as well, and the test updated to use helpers from libtest for working with tunables and xattrs.

Done.

freqlabs commented on cmd/zdb/zdb_il.c:
```diff
@@ -266,6 +266,32 @@ zil_prt_rec_setattr(zilog_t *zilog, int txtype, const void *arg)
+	if (lr->lr_size == 0) {
+		(void) printf("%sXAT_VALUE  NULL\n", tab_prefix);
+	} else {
+		printf("%sXAT_VALUE  ", tab_prefix);
```
```suggestion
		(void) printf("%sXAT_VALUE  ", tab_prefix);
```

jsai20 replied:
Done

freqlabs commented on cmd/zdb/zdb_il.c:
```diff
@@ -266,6 +266,32 @@ zil_prt_rec_setattr(zilog_t *zilog, int txtype, const void *arg)
+		printf("%sXAT_VALUE  ", tab_prefix);
+		val = name + (strlen(name) + 1);
+		for (i = 0; i < lr->lr_size; i++) {
+			printf("%c",  *val);
```
```suggestion
			(void) printf("%c",  *val);
```

jsai20 replied:
Done

freqlabs commented on module/os/freebsd/zfs/zfs_vnops_os.c:
```diff
@@ -5626,7 +5626,7 @@ zfs_setextattr_sa(struct vop_setextattr_args *ap, const char *attrname)
 	kmem_free(buf, entry_size);
 	if (error == 0)
-		error = zfs_sa_set_xattr(zp);
+		error = zfs_sa_set_xattr(zp, attrname, buf, entry_size);
```
`buf` was just freed, let's move that to after this last use.

jsai20 replied:
Thanks. Done.

freqlabs commented on module/zcommon/zfeature_common.c:
```diff
@@ -598,6 +598,18 @@ zpool_feature_init(void)
+	};
+	zfeature_register(SPA_FEATURE_ZILSAXATTR,
+	    "org.zfsonlinux:zilsaxattr", "zilsaxattr",
+	    "Zil sa xattr.",
```
Description too terse IMO.

jsai20 replied:
Done.

freqlabs commented on module/zfs/zfs_replay.c:
```diff
@@ -868,6 +871,89 @@ zfs_replay_setattr(void *arg1, void *arg2, boolean_t byteswap)
+
+	ASSERT(spa_feature_is_active(zfsvfs->z_os->os_spa,
+	    SPA_FEATURE_ZILSAXATTR));
+	if (byteswap) {
```
nit: extra braces inconsistent with the rest of this function

jsai20 replied:
Done

freqlabs commented on module/zfs/zfs_sa.c:
```diff
@@ -244,6 +249,11 @@ zfs_sa_set_xattr(znode_t *zp)
 
+	zilog = zfsvfs->z_log;
+	if (spa_feature_is_active(zfsvfs->z_os->os_spa,
+	    SPA_FEATURE_ZILSAXATTR) && zfs_zil_saxattr_enabled) {
```
nit: extra braces

jsai20 replied:
Done

freqlabs commented on module/zfs/zfs_sa.c:
```diff
@@ -256,6 +266,10 @@ zfs_sa_set_xattr(znode_t *zp)
 		sa_bulk_attr_t bulk[2];
 		uint64_t ctime[2];
 
+		if (logsaxattr) {
```
nit: extra braces

jsai20 replied:
Done

freqlabs commented on module/zfs/zfs_sa.c:
```diff
@@ -433,6 +449,9 @@ zfs_sa_upgrade_txholds(dmu_tx_t *tx, znode_t *zp)
 }
 
+ZFS_MODULE_PARAM(zfs, zfs_, zil_saxattr_enabled, INT, ZMOD_RW,
+	"Enable xattr=sa extended attribute logging in zil.");
```
This looks like a better description to use for the feature flag.

jsai20 replied:
Changed feature flag description to "Support for xattr=sa extended attribute logging in zil."

freqlabs commented:
@jsai20 Can you please rebase this to resolve the module conflict?

jsai20 commented:
> @jsai20 Can you please rebase this to resolve the module conflict?

Done

ahrens commented on module/zcommon/zfeature_common.c:
```diff
@@ -598,6 +598,18 @@ zpool_feature_init(void)
+
+	{
+	static const spa_feature_t zilsaxattr_deps[] = {
+		SPA_FEATURE_EXTENSIBLE_DATASET,
```
[nit] does it really need EXTENSIBLE_DATASET?

jsai20 replied:
Doesn't need. would remove it.

jsai20 replied:
Uploaded new patch to activate feature on zil_create and deactivate during zil_destroy.  It needs per dataset feature flag "ZFEATURE_FLAG_PER_DATASET", which needs EXTENSIBLE_DATASET. So, would keep it.

ahrens commented on module/zcommon/zfeature_common.c:
```diff
@@ -598,6 +598,18 @@ zpool_feature_init(void)
+		SPA_FEATURE_NONE
+	};
+	zfeature_register(SPA_FEATURE_ZILSAXATTR,
+	    "org.zfsonlinux:zilsaxattr", "zilsaxattr",
```
```suggestion
	    "org.openzfs:zilsaxattr", "zilsaxattr",
```

ahrens replied:
Please add the new feature to the zpool-features manpage

jsai20 replied:
Ok

jsai20 replied:
Done

ahrens commented:
@freqlabs Could you take a final look at this?

ahrens commented:
If possible, it would be nice if we could decrement the feature flag when the ZIL is destroyed (e.g. on clean unmount).

freqlabs commented:
> If possible, it would be nice if we could decrement the feature flag when the ZIL is destroyed (e.g. on clean unmount).

So not ACTIVATE_ON_ENABLE? I agree that would be nice.

jsai20 commented on module/zcommon/zfeature_common.c:
```diff
@@ -598,6 +598,18 @@ zpool_feature_init(void)
+		SPA_FEATURE_NONE
+	};
+	zfeature_register(SPA_FEATURE_ZILSAXATTR,
+	    "org.zfsonlinux:zilsaxattr", "zilsaxattr",
```
"log" was bit generic term/word, which could be referred for other logging in system too. Just I just used "zilsaxattr" to be more specific as it stands for "zfs intent log sa xattr".
But If you still, think "logsaxattr" would be better. I can change it.  If do rename here, then would require renaming couple of other places too for function names and variables.

behlendorf replied:
It is a bit generic, but it's also the user visible terminology which is used (e.g `zpool add <pool> log <dev>`), and it aligns with the internal function names, `zfs_log_setsaxattr()`.  So in that regard I think it's pretty clear.  That said, I'm fine with either name so if you'd prefer to keep `zilsaxattr` that's fine.

jsai20 replied:
Keeping "zilsaxattr".

jsai20 commented:
> > If possible, it would be nice if we could decrement the feature flag when the ZIL is destroyed (e.g. on clean unmount).
> 
> So not ACTIVATE_ON_ENABLE? I agree that would be nice.

I would explore the needed change with respect to this.

jsai20 commented on module/zfs/zfs_replay.c:
```diff
@@ -868,6 +871,88 @@ zfs_replay_setattr(void *arg1, void *arg2, boolean_t byteswap)
+	} else {
+		value = name + strlen(name) + 1;
+		/* Limited to 32k to keep nvpair memory allocations small */
+		if (size > DXATTR_MAX_ENTRY_SIZE) {
```
Assert/VERIFY (crash) on BAD log entry could result into repeated panic loops. So, we should avoid ASSERT/VERIFY here.
I think, probably, its better to keep the checks, so that for any reasons Bad log entry found, than better to error out.

behlendorf replied:
The log entries are checksummed so the data should be valid, but I agree there's nothing wrong with being extra careful here and keeping the extra checks.

jsai20 replied:
Ok. Keeping checks.

jsai20 commented on module/zfs/zfs_replay.c:
```diff
@@ -868,6 +871,88 @@ zfs_replay_setattr(void *arg1, void *arg2, boolean_t byteswap)
+	rw_enter(&zp->z_xattr_lock, RW_WRITER);
+	mutex_enter(&zp->z_lock);
+	if (zp->z_xattr_cached == NULL)
+		error = -zfs_sa_get_xattr(zp);
```
Done

jsai20 commented on module/zfs/zfs_replay.c:
```diff
@@ -868,6 +871,88 @@ zfs_replay_setattr(void *arg1, void *arg2, boolean_t byteswap)
+	name = (char *)(lr + 1);
+	if (size == 0) {
+		value = NULL;
+		error = -nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
```
Done

jsai20 commented:
> > > If possible, it would be nice if we could decrement the feature flag when the ZIL is destroyed (e.g. on clean unmount).
> > 
> > 
> > So not ACTIVATE_ON_ENABLE? I agree that would be nice.
> 
> I would explore the needed change with respect to this.

Uploaded new patch adding code to activate "zilsaxattr" feature on zil_create, if its enabled and deactivates it during zil_destroy, if its active.

problame commented on module/zfs/zil.c:
```diff
@@ -723,9 +725,39 @@ zil_create(zilog_t *zilog)
 	if (tx != NULL) {
+		/*
+		 * If "zilsaxattr" feature is enabled on zpool, then activate
+		 * it.
```
`then activate it now when we're creating the ZIL chain. We can't wait with this until we write the first xattr log record because we need to wait for the feature activation to sync out.`

jsai20 replied:
Current patch seems working. So, would discuss offline, to understand bit more on what change is necessary here.

jsai20 replied:
Updated comment as per discussion.

problame commented on module/zfs/zfs_sa.c:
```diff
@@ -264,6 +278,8 @@ zfs_sa_set_xattr(znode_t *zp)
 		VERIFY0(sa_bulk_update(zp->z_sa_hdl, bulk, count, tx));
 
 		dmu_tx_commit(tx);
+		if (logsaxattr && zfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS)
```
Why do we only `zil_commit` for `ZFS_SYNC_ALWAYS`?

I guess because sync semantics of xattrs are not defined? I feel like this should be a separate filesystem property, like, `xattr_sync=on|off`.

Such a property would also make the `zfs_zil_saxattr_enabled` tunable obsolete (if `xattr_sync=off`, act as if the tunable was off).

jsai20 replied:
Yes. Specific sync semantics are not defined for xattr's.
In case of xattr=dir,  zil_commit() is done either when ZFS_SYNC_ALWAYS true on objset OR file opened with (O_SYNC | O_DSYNC).  As here, we only have control to check on ZFS_SYNC_ALWAYS for now, so doing that only.
In future, it can possibly be extended to do have sync control at file level similar to xattr=dir.

problame commented on module/zfs/zil.c:
```diff
@@ -723,9 +725,39 @@ zil_create(zilog_t *zilog)
+	} else {
+		if (spa_feature_is_enabled(zilog->zl_spa,
+		    SPA_FEATURE_ZILSAXATTR) &&
+		    !dsl_dataset_feature_is_active(ds,
```
Regarding our earlier (offline) question about when it's safe to check for feature enablement/activation status.

`spa_feature_is_enabled` is safe because it used the `spa_feat_refcount_cache`, which is updated atomically from `feature_sync`.
Technically, `spa_feature_is_enabled` returns `B_TRUE` a little early because `spa_feat_refcount_cache` is updated _before_ the txg that contains the feature enablement change has finished syncing out.
But oh well, any observer of the `B_TRUE` value can only affect changes in later txgs.
So everything is fine.

Now, regarding `spa_feature_is_active`: that uses the same `spa_feat_refcount_cache` infrastructure.
But some features can flip back from active to enabled.
So, observers that make changes based on `spa_feature_is_active == B_TRUE` must be certain that the feature will still be active when the change syncs out.

And finally, regarding `dsl_dataset_feature_is_active`: it seems to me like it's only safe to use that function from syncing context. The reason is that `ds_feature` updates are only made in syncing context and are not guarded by locks.
That being said, there are a bunch of places in the code base where we call `dsl_dataset_feature_is_active` outside of syncing context.
I guess that, while its technically a data race, there won't be any partial reads on Intel platforms, so we're safe.
But this is probably something that should be fixed.

jsai20 replied:
Current patch seems working :). So, would avoid making any more changes until necessary.

jsai20 replied:
As discussed, keeping it. Added comment explaining need of using dsl_dataset_feature_is_active();

problame commented on module/zfs/zil.c:
```diff
@@ -723,9 +725,39 @@ zil_create(zilog_t *zilog)
+		    SPA_FEATURE_ZILSAXATTR) &&
+		    !dsl_dataset_feature_is_active(ds,
+		    SPA_FEATURE_ZILSAXATTR)) {
+			tx = dmu_tx_create(zilog->zl_os);
```
Notes from conversation earlier today: this branch covers the case where we enable the feature on a zpool that has existing ZIL headers.

jsai20 replied:
yes

problame reviewed (commented):
Notes from our offline conversation earlier today, plus some remarks.

jsai20 commented:
@freqlabs @behlendorf Please post, if you have any more comments.

problame reviewed (changes requested):
I gave this another review pass.
The feature activation in `zil_create` is insufficient in the following case:

1. create pool with feature enabled
2. create dataset with xattr=sa
3. do a few sync writes to start a ZIL chain
4. set an xattr user.foo=bar
5. crash immediately

In (4), the ZIL will have already been created in `zl_lwb_list` not empty, which means `zfs_create` won't be called, and hence the feature won't be activated.

I think we need to check `dsl_dataset_feature_is_active` from `zil_process_commit_list`, and if it isn't, activate the feature, and wait until the txg has synced before proceeding with writing log records.
Perhaps we can share some code with `zil_commit_writer_stall`.

nabijaczleweli commented on man/man7/zpool-features.7:
```diff
@@ -836,6 +836,15 @@ once all filesystems that have ever had their
 are destroyed.
+.
+.feature org.openzfs zilsaxattr no extensible_dataset
+This feature enables xattr=sa extended attribute logging in ZIL.
+.Pp
+This feature becomes
+.Sy active
+when ZIL for a dataset is created and will be returned to the
+.Sy enabled
+state when ZIL is destroyed for all datasets that use this feature.
```
```
This feature enables
.Sy xattr Ns = Ns Sy sa
extended attribute logging in the ZIL.
.Pp
This feature becomes
.Sy active
when a ZIL is created for at least one dataset and will be returned to the
.Sy enabled
state when it is destroyed for all datasets that use this feature.
```

jsai20 replied:
Done

nabijaczleweli commented on man/man4/zfs.4:
```diff
@@ -2092,6 +2092,10 @@ Limit SLOG write size per commit executed with synchronous priority.
 Any writes above that will be executed with lower (asynchronous) priority
 to limit potential SLOG device abuse by single active ZIL writer.
 .
+.It Sy zfs_zil_saxattr_enabled Ns = Ns Sy 1 Pq int
+Enable xattr=sa extended attribute logging in ZIL.
+Setting value to 0 disables xattr=sa logging in ZIL.
```
```
.It Sy zfs_zil_saxattr_enabled Ns = Ns Sy 1 Ns | Ns 0 Pq int
Enable/disable
.Sy xattr Ns = Ns Sy sa
extended attribute logging in the ZIL.
```

jsai20 replied:
Done

nabijaczleweli commented on module/zfs/zfs_sa.c:
```diff
@@ -69,6 +71,8 @@ sa_attr_reg_t zfs_attr_table[ZPL_END+1] = {
 	{NULL, 0, 0, 0}
 };
 
+int zfs_zil_saxattr_enabled = 1;
```
No reason this can't be `static` I think?

jsai20 replied:
Done

jsai20 commented on module/zfs/zfs_sa.c:
```diff
@@ -244,6 +249,11 @@ zfs_sa_set_xattr(znode_t *zp)
 		goto out_free;
 
+	zilog = zfsvfs->z_log;
+	if (spa_feature_is_active(zfsvfs->z_os->os_spa,
```
Done

jsai20 commented:
> I gave this another review pass. The feature activation in `zil_create` is insufficient in the following case:
> 
>     1. create pool with feature enabled
> 
>     2. create dataset with xattr=sa
> 
>     3. do a few sync writes to start a ZIL chain
> 
>     4. set an xattr user.foo=bar
> 
>     5. crash immediately
> 
> 
> In (4), the ZIL will have already been created in `zl_lwb_list` not empty, which means `zfs_create` won't be called, and hence the feature won't be activated.
> 
> I think we need to check `dsl_dataset_feature_is_active` from `zil_process_commit_list`, and if it isn't, activate the feature, and wait until the txg has synced before proceeding with writing log records. Perhaps we can share some code with `zil_commit_writer_stall`.

Added code to activate feature in zil_process_commit_list() for the case where zl_lwb_list is not empty, feature is enabled but not activated.

problame reviewed (commented):
So, to summarize, the feature activation now works like this:

```
zil_commit()
  zil_process_commit_list()
    if lwb == NULL // first zil_commit since zil_open
      zil_create()
        if no log block pointer in ZIL header:
          if feature enabled and not active:
	    // CASE 1
            enable, COALESCE txg wait with dmu_tx that allocated the log block
         else // log block was allocated earlier than this zil_open
          if feature enabled and not active:
	    // CASE 2
            enable, EXPLICIT txg wait
    else // already have an in-DRAM LWB
      if feature enabled and not active:
        // this happens when we enable the feature after zil_create
	// CASE 3
        enable, EXPLICIT txg wait
```

And feature deactivation happens from zil_sync(), initated from zil_destroy().

Correct?

nabijaczleweli commented on cmd/zdb/zdb_il.c:
```diff
@@ -265,6 +265,32 @@ zil_prt_rec_setattr(zilog_t *zilog, int txtype, const void *arg)
+	}
+}
+
+/* ARGSUSED */
```
Extraneous ARGSUSED

jsai20 replied:
Done

jsai20 commented:
> So, to summarize, the feature activation now works like this:
> 
> ```
> zil_commit()
>   zil_process_commit_list()
>     if lwb == NULL // first zil_commit since zil_open
>       zil_create()
>         if no log block pointer in ZIL header:
>           if feature enabled and not active:
> 	    // CASE 1
>             enable, COALESCE txg wait with dmu_tx that allocated the log block
>          else // log block was allocated earlier than this zil_open
>           if feature enabled and not active:
> 	    // CASE 2
>             enable, EXPLICIT txg wait
>     else // already have an in-DRAM LWB
>       if feature enabled and not active:
>         // this happens when we enable the feature after zil_create
> 	// CASE 3
>         enable, EXPLICIT txg wait
> ```
> 
> And feature deactivation happens from zil_sync(), initated from zil_destroy().
> 
> Correct?

yes

behlendorf reviewed (approved):
Just a few minor nits, looks good.

jsai20 commented on man/man7/zpool-features.7:
```diff
@@ -832,6 +832,17 @@ once all filesystems that have ever had their
 .Sy zstd
 are destroyed.
+.
+.feature org.openzfs zilsaxattr no extensible_dataset
```
Done

jsai20 commented on module/zcommon/zfeature_common.c:
```diff
@@ -695,7 +695,20 @@ zpool_feature_init(void)
+	}
+
 	zfs_mod_list_supported_free(sfeatures);
+
```
Done

jsai20 commented on module/zfs/zfs_replay.c:
```diff
@@ -868,6 +871,88 @@ zfs_replay_setattr(void *arg1, void *arg2, boolean_t byteswap)
+		zp->z_xattr_cached = NULL;
+	}
+
+	ASSERT3S(error, <=, 0);
```
Done

jsai20 commented on module/zfs/zil.c:
```diff
@@ -663,6 +663,35 @@ zilog_is_dirty(zilog_t *zilog)
+ * Its called in zil_commit context (zil_process_commit_list()/zil_creat()).
+ * It activates SPA_FEATURE_ZILSAXATTR feature, if its enabled.
+ */
+static inline void
```
Done

behlendorf commented:
@ahrens @freqlabs @problame would you mind taking a final look at this.

problame reviewed (changes requested):
I added a bunch of suggestions for better comments / docs.

Also, I think we should reflect some of the complexities of feature activation in the commit message.
Suggestion for rephrased commit message:

```
log xattr=sa create/remove/update to ZIL

As such, there are no specific synchronous semantics defined for
the xattrs. But for xattr=on, it does log to ZIL and zil_commit() is
done, if sync=always is set on dataset. This provides sync semantics
for xattr=on with sync=always set on dataset.

For the xattr=sa implementation, it doesn't log to ZIL, so, even with
sync=always, xattrs are not guaranteed to be synced before xattr call
returns to caller. So, xattr can be lost if system crash happens, before
txg carrying xattr transaction is synced.

This change adds xattr=sa logging to ZIL on xattr create/remove/update
and xattrs are synced to ZIL (zil_commit() done) for sync=always.
This makes xattr=sa behavior similar to xattr=on.

Implementation notes:
The actual logging is fairly straight-forward and does not warrant
additional explanation.
However, it has been 14 years since we last added new TX types
to the ZIL [1], hence this is the first time we do it after the introduction
of zpool features. Therefore, here is an overview of the feature
activation and deactivation workflow:

1. The feature must be enabled. Otherwise, we don't log the new
    record type. This ensures compatibility with older software.
2. The feature is activated per-dataset, since the ZIL is per-dataset.
3. If the feature is enabled, any append to the ZIL chain will activate
    the feature for the dataset. Likewise for starting a new ZIL chain.
4. A dataset that doesn't have a ZIL chain has the feature deactivated.

We ensure (3) by activating on the first zil_commit() after the feature
was enabled. Since activating the features requires waiting for txg sync,
the first zil_commit() after enabling the feature will be slower than usual.
The downside is that this is really a conservative approximation: even
if we never append a `TX_SETSAXATTR` to the ZIL chain, we pay
the penalty for feature activation. The upside is that the user is in
control of when we pay the penalty, i.e., upon enabling the feature.

We ensure (4) by hooking into zil_sync(), where ZIL destroy actually
happens.

One more piece on feature activation, since it's spread across
multiple functions:

zil_commit()
  zil_process_commit_list()
    if lwb == NULL // first zil_commit since zil_open
      zil_create()
        if no log block pointer in ZIL header:
          if feature enabled and not active:
	    // CASE 1
            enable, COALESCE txg wait with dmu_tx that allocated the log block
         else // log block was allocated earlier than this zil_open
          if feature enabled and not active:
	    // CASE 2
            enable, EXPLICIT txg wait
    else // already have an in-DRAM LWB
      if feature enabled and not active:
        // this happens when we enable the feature after zil_create
	// CASE 3
        enable, EXPLICIT txg wait

[1] https://github.com/illumos/illumos-gate/commit/da6c28aaf62fa55f0fdb8004aa40f88f23bf53f0
```

jsai20 commented:
> I added a bunch of suggestions for better comments / docs.
> 
> Also, I think we should reflect some of the complexities of feature activation in the commit message. Suggestion for rephrased commit message:
> 
> ```
> log xattr=sa create/remove/update to ZIL
> 
> As such, there are no specific synchronous semantics defined for
> the xattrs. But for xattr=on, it does log to ZIL and zil_commit() is
> done, if sync=always is set on dataset. This provides sync semantics
> for xattr=on with sync=always set on dataset.
> 
> For the xattr=sa implementation, it doesn't log to ZIL, so, even with
> sync=always, xattrs are not guaranteed to be synced before xattr call
> returns to caller. So, xattr can be lost if system crash happens, before
> txg carrying xattr transaction is synced.
> 
> This change adds xattr=sa logging to ZIL on xattr create/remove/update
> and xattrs are synced to ZIL (zil_commit() done) for sync=always.
> This makes xattr=sa behavior similar to xattr=on.
> 
> Implementation notes:
> The actual logging is fairly straight-forward and does not warrant
> additional explanation.
> However, it has been 14 years since we last added new TX types
> to the ZIL [1], hence this is the first time we do it after the introduction
> of zpool features. Therefore, here is an overview of the feature
> activation and deactivation workflow:
> 
> 1. The feature must be enabled. Otherwise, we don't log the new
>     record type. This ensures compatibility with older software.
> 2. The feature is activated per-dataset, since the ZIL is per-dataset.
> 3. If the feature is enabled, any append to the ZIL chain will activate
>     the feature for the dataset. Likewise for starting a new ZIL chain.
> 4. A dataset that doesn't have a ZIL chain has the feature deactivated.
> 
> We ensure (3) by activating on the first zil_commit() after the feature
> was enabled. Since activating the features requires waiting for txg sync,
> the first zil_commit() after enabling the feature will be slower than usual.
> The downside is that this is really a conservative approximation: even
> if we never append a `TX_SETSAXATTR` to the ZIL chain, we pay
> the penalty for feature activation. The upside is that the user is in
> control of when we pay the penalty, i.e., upon enabling the feature.
> 
> We ensure (4) by hooking into zil_sync(), where ZIL destroy actually
> happens.
> 
> One more piece on feature activation, since it's spread across
> multiple functions:
> 
> zil_commit()
>   zil_process_commit_list()
>     if lwb == NULL // first zil_commit since zil_open
>       zil_create()
>         if no log block pointer in ZIL header:
>           if feature enabled and not active:
> 	    // CASE 1
>             enable, COALESCE txg wait with dmu_tx that allocated the log block
>          else // log block was allocated earlier than this zil_open
>           if feature enabled and not active:
> 	    // CASE 2
>             enable, EXPLICIT txg wait
>     else // already have an in-DRAM LWB
>       if feature enabled and not active:
>         // this happens when we enable the feature after zil_create
> 	// CASE 3
>         enable, EXPLICIT txg wait
> 
> [1] https://github.com/illumos/illumos-gate/commit/da6c28aaf62fa55f0fdb8004aa40f88f23bf53f0
> ```

Replaced commit message as suggested.

jsai20 commented on man/man4/zfs.4:
```diff
@@ -2098,6 +2098,11 @@ Limit SLOG write size per commit executed with synchronous priority.
+.It Sy zfs_zil_saxattr_enabled Ns = Ns Sy 1 Ns | Ns 0 Pq int
+Enable/disable
+.Sy xattr Ns = Ns Sy sa
+extended attribute logging in the ZIL.
```
Renamed to tunable to "zfs_zil_saxattr" and added description as suggested.

jsai20 commented on module/zfs/zil.c:
```diff
@@ -663,6 +663,35 @@ zilog_is_dirty(zilog_t *zilog)
 }
 
+/*
+ * Its called in zil_commit context (zil_process_commit_list()/zil_creat()).
```
Done

jsai20 commented on module/zfs/zil.c:
```diff
@@ -724,9 +755,33 @@ zil_create(zilog_t *zilog)
+		 * Check for dsl dataset feature is not active to avoid
+		 * txg_wait_synced() on every zil_commit.
+		 */
+		zil_commit_activate_saxattr_feature(zilog);
```
Done

nabijaczleweli commented on man/man4/zfs.4:
```diff
@@ -2098,6 +2098,13 @@ Limit SLOG write size per commit executed with synchronous priority.
+.It Sy zfs_zil_saxattr Ns = Ns Sy 1 Ns | Ns 0 Pq int
+Setting this tunable to zero disables ZIL logging of new
+.Sy xattr Ns = Ns Sy sa
+records if feature org.openzfs:zilsaxattr is enabled on the pool. This would
+only be necessary to work around bugs in the ZIL logging or replay code for this
+record type. The tunable has no effect if the zpool feature org.openzfs:zilsaxattr is disabled.
```
```
records if the
.Sy org.openzfs:zilsaxattr
feature is enabled on the pool.
This would only ... type.
The tunable has no effect if the feature is disabled.
```

jsai20 replied:
Done. Thanks!

jsai20 commented on module/zfs/zil.c:
```diff
@@ -3092,6 +3153,14 @@ zil_sync(zilog_t *zilog, dmu_tx_t *tx)
 			zh->zh_log = blk;
+		} else {
+			/*
+			 * Deactivate "zilsaxattr" feature, if it is active.
```
Done

jsai20 commented on man/man7/zpool-features.7:
```diff
@@ -778,6 +778,17 @@ by user and group.
+.feature org.openzfs zilsaxattr no extensible_dataset
+This feature enables
+.Sy xattr Ns = Ns Sy sa
+extended attribute logging in the ZIL.
```
Updated as suggested.

nabijaczleweli commented on man/man7/zpool-features.7:
```diff
@@ -778,6 +778,20 @@ by user and group.
+This feature enables
+.Sy xattr Ns = Ns Sy sa
+extended attribute logging in the ZIL. If enabled, extended attribute changes
+(both xattr=dir and xattr=sa) are guaranteed to be durable if either the dataset
+had sync=always set at the time the changes were made, or sync() is called on
```
```
.Pq both Sy xattrdir Ns = Ns Sy dir No and Sy xattr Ns = Ns Sy sa
are guaranteed .. had
.Sy sync Ns = Ns Sy always
set .. made, or
.Fn sync
is called on
```

jsai20 replied:
Done. Thanks.

freqlabs commented on module/zfs/zfs_replay.c:
```diff
@@ -868,6 +871,86 @@ zfs_replay_setattr(void *arg1, void *arg2, boolean_t byteswap)
+	 * will be reconstructed from the ARC when next accessed.
+	 */
+	if (error == 0)
+		error = zfs_sa_set_xattr(zp, name, value, size);
```
Doesn't this end up logging the operation we just replayed again? Am I missing something?

freqlabs replied:
Ah I see what I missed, the replaying check in the log function.

scineram commented on module/zfs/zil.c:
```diff
@@ -724,9 +757,31 @@ zil_create(zilog_t *zilog)
+		zil_commit_activate_saxattr_feature(zilog);
 	}
+	IMPLY(spa_feature_is_enabled(zilog->zl_spa, SPA_FEATURE_ZILSAXATTR),
+	    dsl_dataset_feature_is_active(ds, SPA_FEATURE_ZILSAXATTR));
```
Is this feature also activated for zvols (unnecessarily)?

jsai20 replied:
Yes. I would add a check to skip activating feature for zvol's.
I would avoid adding code around activating feature only when dataset, if it has xattr=sa, as it makes overall things more complicated.  As such, If any consumers not using xattr=sa, then they can choose to not enable this feature.

behlendorf reviewed (approved):
As of today, we enabled some additional compiler warning.  If you could rebase this on the latet master when addressing th last few comments I posted we can verify it still builds cleanly.

jsai20 commented on cmd/zdb/zdb_il.c:
```diff
@@ -265,6 +265,31 @@ zil_prt_rec_setattr(zilog_t *zilog, int txtype, const void *arg)
 
+static void
+zil_prt_rec_setsaxattr(zilog_t *zilog, int txtype, const void *arg)
+{
```
Done

jsai20 commented on cmd/zdb/zdb_il.c:
```diff
@@ -265,6 +265,31 @@ zil_prt_rec_setattr(zilog_t *zilog, int txtype, const void *arg)
+	char	*val;
+	int	i;
+
+	name = (char *)(lr + 1);
```
Done

jsai20 commented on cmd/zdb/zdb_il.c:
```diff
@@ -265,6 +265,31 @@ zil_prt_rec_setattr(zilog_t *zilog, int txtype, const void *arg)
+		(void) printf("%sXAT_VALUE  NULL\n", tab_prefix);
+	} else {
+		(void) printf("%sXAT_VALUE  ", tab_prefix);
+		val = name + (strlen(name) + 1);
```
Done

jsai20 commented on man/man7/zpool-features.7:
```diff
@@ -778,6 +778,24 @@ by user and group.
 \*[instant-never]
 \*[remount-upgrade]
 .
+.feature org.openzfs zilsaxattr no extensible_dataset
```
Done

jsai20 commented on module/zfs/zfs_replay.c:
```diff
@@ -47,6 +47,9 @@
 #include <sys/atomic.h>
 #include <sys/cred.h>
 #include <sys/zpl.h>
+#include <sys/dmu.h>
```
Done

behlendorf commented:
@jsai20 thanks for iterating with us on this, merged!

jsai20 commented:
> @jsai20 thanks for iterating with us on this, merged!

Thanks @behlendorf @ahrens @freqlabs @problame @nabijaczleweli @scineram and othre reviewers.  Thanks much for refining it!
